statistics/handle/lock_stats_handler.go

```diff
@@ -49,161 +29,50 @@ var (
-	if err != nil {
-		logutil.BgLogger().Error("error occurred when insert mysql.stats_table_locked", zap.String("category", "stats"), zap.Error(err))
-		return err
-	}
-	return nil
+	h.mu.RLock()
+	defer h.mu.RUnlock()
+	return lockstats.AddLockedTables(h.mu.ctx.(sqlexec.SQLExecutor), tids, pids, tables)
 }
 
-// LoadLockedTables load locked tables from store
-func (h *Handle) LoadLockedTables() error {
+// RemoveLockedTables remove tables from table locked array.
+// - tids: table ids of which will be unlocked.
+// - pids: partition ids of which will be unlocked.
+// - tables: table names of which will be unlocked.
+// Return the message of skipped tables and error.
+func (h *Handle) RemoveLockedTables(tids []int64, pids []int64, tables []*ast.TableName) (string, error) {
 	h.mu.Lock()
 	defer h.mu.Unlock()
-	ctx := kv.WithInternalSourceType(context.Background(), kv.InternalTxnStats)
-	exec := h.mu.ctx.(sqlexec.SQLExecutor)
-
-	tableLocked, err := loadLockedTables(ctx, exec, maxChunkSize)
-	if err != nil {
-		return err
-	}
 
-	// Update all locked tables.
-	h.tableLocked = tableLocked
-
-	return nil
+	return lockstats.RemoveLockedTables(h.mu.ctx.(sqlexec.SQLExecutor), tids, pids, tables)
 }
 
 // IsTableLocked check whether table is locked in handle with Handle.Mutex
-func (h *Handle) IsTableLocked(tableID int64) bool {
+func (h *Handle) IsTableLocked(tableID int64) (bool, error) {
 	h.mu.RLock()
 	defer h.mu.RUnlock()
 	return h.isTableLockedWithoutLock(tableID)
 }
 
-// isTableLockedWithoutLock check whether table is locked in handle without Handle.Mutex
-func (h *Handle) isTableLockedWithoutLock(tableID int64) bool {
-	return isTableLocked(h.tableLocked, tableID)
-}
-
-// isTableLocked check whether table is locked
-func isTableLocked(tableLocked []int64, tableID int64) bool {
-	return lockTableIndexOf(tableLocked, tableID) > -1
-}
+// loadLockedTablesWithoutLock load locked tables from store without Handle.Mutex.
+func (h *Handle) loadLockedTablesWithoutLock() ([]int64, error) {
+	ctx := kv.WithInternalSourceType(context.Background(), kv.InternalTxnStats)
+	exec := h.mu.ctx.(sqlexec.SQLExecutor)
 
-// GetTableLockedAndClearForTest for unit test only
-func (h *Handle) GetTableLockedAndClearForTest() []int64 {
-	tableLocked := h.tableLocked
-	h.tableLocked = make([]int64, 0)
-	return tableLocked
+	return lockstats.LoadLockedTables(ctx, exec)
 }
 
-func loadLockedTables(ctx context.Context, exec sqlexec.SQLExecutor, maxChunkSize int) ([]int64, error) {
-	recordSet, err := exec.ExecuteInternal(ctx, "SELECT table_id FROM mysql.stats_table_locked")
-	if err != nil {
-		return nil, err
-	}
-	rows, err := sqlexec.DrainRecordSet(ctx, recordSet, maxChunkSize)
+// isTableLockedWithoutLock check whether table is locked in handle without Handle.Mutex
+func (h *Handle) isTableLockedWithoutLock(tableID int64) (bool, error) {
```

Comment:
`h.mu` lock's granularity is too coarse and not reentrant, which brings lots of risks to us.
For example, someone may don't know whether he should use `h.IsTableLocked` or `h.IsTableLockedWithoutLock`, and once they use the wrong method, we'll get a terrible DATA RACE or DEADLOCK problem.
How about adding a TODO here to notify us to replace this lock with a reentrant lock in the future?